### PR TITLE
Fix performance issue with NameError#missing_name on ruby >= v2.3.0.

### PR DESCRIPTION
**Pull request by ayanko**
Fix performance issue on ruby with `did_you_mean` gem.

### Summary

Since ruby v2.3.0 `did_you_mean` gem shipped and ENABLED by default.
It patches NameError#message with spell corrections which are SLOW.
We should use `original_message` instead.

### Another options

* gem uninstall did_you_mean
* ruby --disable-did-you-mean -e ...
* RUBYOPT=--disable-did-you-mean ruby ...
* add RUBYOPT=--disable-did-you-mean to your bash/dash/foo shell rc

That's very bad that this fancy gem is enabled by default (even in production environment).
But anyway if it enabled rails should NOT be slow.

### Benchmark script

```ruby
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"
  gem "rails", path: ENV.fetch('RAILS_REPO')
end

require "active_support"
require "active_support/core_ext/name_error"
require "benchmark"


class Foo
end

puts "-"*80
p RUBY_VERSION
p NameError.ancestors

Benchmark.bmbm do |x|
  x.report 'NameError#missing_message' do
    100.times do
      begin
        Foo1
      rescue NameError => e
        e.missing_name
      end
    end
  end
end
```

### Original version

`$ RAILS_REPO=rails/rails ruby bench.rb`

```
--------------------------------------------------------------------------------
"2.4.1"
[DidYouMean::Correctable, NameError, StandardError, Exception, ActiveSupport::ToJsonWithActiveSupportEncoder, Object, JSON::Ext::Generator::GeneratorMethods::Object, ActiveSupport::Tryable, Kernel, BasicObject]
Rehearsal -------------------------------------------------------------
NameError#missing_message   0.100000   0.000000   0.100000 (  0.099724)
---------------------------------------------------- total: 0.100000sec

                                user     system      total        real
NameError#missing_message   0.100000   0.000000   0.100000 (  0.096193)
```

### Optimized version

`RAILS_REPO=ayanko/rails ruby bench.rb`

```
--------------------------------------------------------------------------------
"2.4.1"
[DidYouMean::Correctable, NameError, StandardError, Exception, ActiveSupport::ToJsonWithActiveSupportEncoder, Object, JSON::Ext::Generator::GeneratorMethods::Object, ActiveSupport::Tryable, Kernel, BasicObject]
Rehearsal -------------------------------------------------------------
NameError#missing_message   0.000000   0.000000   0.000000 (  0.001005)
---------------------------------------------------- total: 0.000000sec

                                user     system      total        real
NameError#missing_message   0.000000   0.000000   0.000000 (  0.000767)
```

### Conclusion

`DidYouMean` brings **serious performance degradation**.

### CODE REVIEW

**Comment by matthewd** [comment]
The change is simple enough, so it seems fine, despite failing to provide a real benchmark. Let's just use a local variable instead of a method, though.

> rails should NOT be slow

> **serious performance degradation**

This is wildly overstating the impact, to the point of ridicule.

A casual glance shows that only one of the *two* callers of this method will even be invoked in production, and that will only occur at boot. (In development, they're both similarly only involved in loading, but with class reloading, that could occur more than once.)

**Comment by ayanko** [comment]
>A casual glance shows that only one of the two callers of this method will even be invoked in production, and that will only occur at boot.

It's not true. There are gems that use this method. For instance draper:
https://github.com/drapergem/draper/blob/v3.0.0/lib/draper/decoratable.rb#L77

I use draper in production and I have general presenter that tries to decorate model first and this use `decorator_class?` method to detect if model can be decorated. Because I don't need decorators for all my models.

So due to this problem (did_you_mean) I have real performance degradation in request.
E.g.: Two  NameError#missing_name costs about **40ms**

**Comment by ayanko** [comment]
> Let's just use a local variable instead of a method, though.

@matthewd DONE

**Comment by matthewd** [comment]
Is there any practical way we can test both paths here?

(It's always good to prevent regressions, but I'm also pretty sure the no-`original_message` path is currently broken.)

**Comment by ayanko** [comment]
> Is there any practical way we can test both paths here?

I don't know elegant way to do it.
But name `original_message` is abstract enough and I guess it claims that is actually non-modified original message :)

**Comment by ayanko** [comment]
Conclusion?

**Comment by matthewd** [comment]
> I'm also pretty sure the no-`original_message` path is currently broken

**Comment by ayanko** [comment]
@matthewd I don't understand your answer. Could you explain?

**Comment by matthewd** [comment]
After this change, `missing_name` will not work when `respond_to?(:original_message)` is false.

File: activesupport/lib/active_support/core_ext/name_error.rb [hunk 1]
```diff
@@ -8,6 +8,11 @@ class NameError
   #   end
   #   # => "HelloWorld"
   def missing_name
+    # Since ruby v2.3.0 `did_you_mean` gem is loaded by default.
+    # It extends NameError#message with spell corrections which are SLOW.
+    # We should use original_message message instead.
+    message = respond_to?(:original_message) ? original_message : message
```
Comment (kamipo) [inline activesupport/lib/active_support/core_ext/name_error.rb]: `message = respond_to?(:original_message) ? original_message : self.message`

**Comment by ayanko** [comment]
You are right. Updated.

**Comment by ayanko** [comment]
@matthewd Conclusion?
